### PR TITLE
Add layouts to the template

### LINKED ISSUE #50: [feature request]  layout feature
Hi, great works on this!!

Providing layout feature just like nuxt.js would be very helpful.
I'm wondering if it is even possible to have such feature in vite.
Do you have any suggestion on how to do this?
Thanks!!

### PR DESCRIPTION
**Pull request by JohnCampionJr**
Closes #50 

Let me know if you'd like any particular changes to the samples or anything.

### CODE REVIEW

File: src/layouts/README.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+## Layouts
+
+Vue components in this dir are used as layouts.
+
+The default.vue is used for all pages unless an alternative is specified in the route meta.
+
+With vite-plugin-pages, you can specify the layout in the page's .vue file like this:
+```html
+<route>
```
Comment (antfu) [inline src/layouts/README.md:9]: I'd love to feature `lang='yaml'` here for a clear code style.

**Review by antfu — COMMENTED** [review]
Overall looks good, I will do some tweaks to the styling later.

I would wish we could have an abstract of `layouts/_setup.ts`, maybe a plugin, or a build-in feature of `vite-plugin-pages` (/cc @hannoeru WDTY?) or find a better name for it. The inlined code is a bit too complicated and verbose for a project template from my personal opinion.

Thanks for the work!

**Comment by JohnCampionJr** [comment]
I was thinking that too.  But it's not strictly a vite plugin.  

Let me start a vitesse-layout plugin repo and take a spin at it that way.

File: src/layouts/_setup.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      component: RouterLayout,
+      children: routes,
+    },
+  ]
+}
+
+export function createRouterLayout(
+  resolve: (layoutName: string) => Promise<Component | { default: Component }>,
+) {
+  return defineComponent({
```
[the remark this answers is not in the document]
Comment (JohnCampionJr) [reply src/layouts/_setup.ts:39]: I don't think so.  beforeRouteEnter is not available with Composition API.

https://next.router.vuejs.org/guide/advanced/composition-api.html

File: src/layouts/_setup.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  VNode,
+  Component,
+  ConcreteComponent,
+  h,
+  defineComponent,
+  shallowReactive,
+} from 'vue'
+import { RouteRecordRaw } from 'vue-router'
+
+function normalizeEsModuleComponent(
```
[the remark this answers is not in the document]
Comment (JohnCampionJr) [reply src/layouts/_setup.ts:12]: You are correct; removing this.

**Comment by JohnCampionJr** [comment]
I broke my branch trying to rebase onto your changes.   Let me start a fresh branch and try another PR

**Comment by antfu** [comment]
> The import may not work if it is sitting in node_modules. Vite is pretty specific about its importing options.

It's not necessary to use `import` exclusively. In the plugin, we can do a glob search and then generate the code to be served by a virtual module, just like what `vite-plugin-pages` is doing. For the config, could simply be an option like

```ts
VitePages({
  layoutDirs: ['./layouts']
})
```

File: src/layouts/_setup.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (JohnCampionJr) [reply src/layouts/_setup.ts:12]: Well, I took out the function and the whole thing broke.

**Comment by JohnCampionJr** [comment]
> It's not necessary to use `import` exclusively. In the plugin, we can do a glob search and then generate the code to be served by a virtual module, just like what `vite-plugin-pages` is doing. 

So just add the code by string with Vite-pages.  That could work.  Should I close this for now?

**Comment by JohnCampionJr** [comment]
Very sorry I messed up this branch and its PR.  I tried to rebase the branch onto the changes made and messed something up.

**Pull request closed without merging**